### PR TITLE
Preprocessing next features

### PR DESCRIPTION
**Pull request by t0maboro**
- [x] columns alignment with a null value
- [x] plot showing null and not null occurrence
- [x] reduce number of labels on plot - dependable on some properties
- [x] remove rows with any null value after user's consent
- [ ] date-time integration from different sources

### CODE REVIEW

File: src/widgets/plots/pie_plot.py [hunk 1]
```diff
@@ -6,7 +7,20 @@ def __init__(self, data):
...
     def plot(self):
-        ax = self.canvas.figure.subplots()
-        counts = self.data.value_counts().to_dict()
-        ax.pie(counts.values(), labels=counts.keys())
-        return self.canvas
+        try:
+            ax = self.canvas.figure.subplots()
+            counts = self.data.value_counts().to_dict()
+            data_size = self.data.size
+            labels = [k if counts[k]/data_size > self.min_pie_plot_label_ratio else '' for k in counts.keys()]
```
Comment (gregori0o) [inline src/widgets/plots/pie_plot.py]: Maybe plot few big pieces and set label for them and other values merge to one piece in circle and set label 'other'. This one piece may be in one color or divide like now.

File: src/widgets/plots/histogram_plot.py [hunk 2]
```diff
@@ -6,6 +7,26 @@ def __init__(self, data):
...
-        ax.hist(self.data)
-        return self.canvas
+        try:
+            ax = self.canvas.figure.subplots()
+            stats = self.data.value_counts().to_dict()
+            labels, values = zip(*stats.items())
+            ax.bar(labels, values, align='center')
+            if len(labels) > self.max_labels_show:
+                self._reduce_labels(ax, len(labels))
+            return self.canvas
```
Comment (gregori0o) [inline src/widgets/plots/histogram_plot.py]: Labels on OX axis have vertically direction.
image

File: src/widgets/plots/null_freq_plot.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        try:
+            ax = self.canvas.figure.subplots()
+            nulls = self.data.value_counts(normalize=True).get('null', 0)
+            not_nulls = 1 - nulls
+            nulls_bar = ax.bar(0.25, nulls, 0.5, label='NULL')
+            not_nulls_bar = ax.bar(0.75, not_nulls, 0.5, label='NOT NULL')
+            ax.legend()
+            ax.bar_label(nulls_bar)
+            ax.bar_label(not_nulls_bar)
+            return self.canvas
```
Comment (gregori0o) [inline src/widgets/plots/null_freq_plot.py]: In this plot columns are very wide and it looks weird. I think that information about None value may be in Estimate missing values section or you should change appearance of this plot, but now I don't have any idea how change this.

OX axis have number labels, could you disable this labels?

**Review by gregori0o — COMMENTED** [review]
Good job!
Only small suggestion.

File: src/engines/preprocessing_engine.py [hunk 4]
```diff
@@ -26,10 +27,16 @@ def create_plot(self, column_name, plot_type):
...
+            case 'Null frequency':
+                plotter = NullFrequencyPlot(column)
         return plotter.plot()
 
-    def _cast_nulls(self):
-        self.state.imported_data = self.state.imported_data.fillna("null")
+    def clean_data(self, op_type):
+        match op_type:
+            case "cast":
+                self.cleaner.cast_nulls("null")
```
Comment (mhawryluk) [inline src/engines/preprocessing_engine.py]: We shouldn't represent an absent value with a "null" string. I suggest we stick with np.NaN.

One of many downsides of using a string is that when we have a column with only numerical values and some empty cells we cannot draw a histogram, as the types are incompatible, whereas NaN values would be automatically ignored.

File: src/widgets/plots/histogram_plot.py [hunk 5]
```diff
@@ -6,6 +7,26 @@ def __init__(self, data):
...
-        ax = self.canvas.figure.subplots()
-        ax.hist(self.data)
-        return self.canvas
+        try:
+            ax = self.canvas.figure.subplots()
+            stats = self.data.value_counts().to_dict()
+            labels, values = zip(*stats.items())
+            ax.bar(labels, values, align='center')
+            if len(labels) > self.max_labels_show:
+                self._reduce_labels(ax, len(labels))
```
Comment (mhawryluk) [inline src/widgets/plots/histogram_plot.py]: For numerical values on that axis it makes it so there are very few labels. len(labels) is typically much bigger than len(ax.xaxis.get_ticklabels()). I believe they're the same thing just for strings.

File: src/widgets/plots/null_freq_plot.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class NullFrequencyPlot(Plot):
+    def __init__(self, data):
+        super().__init__(data)
+
+    def plot(self):
+        try:
+            ax = self.canvas.figure.subplots()
+            nulls = self.data.value_counts(normalize=True).get('null', 0)
```
Comment (mhawryluk) [inline src/widgets/plots/null_freq_plot.py]: For when the ""null"" is replaced it should probably look something like that:
```suggestion
            nulls = self.data.isna().sum()/len(self.data)
```

File: src/widgets/plots/null_freq_plot.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+            not_nulls = 1 - nulls
+            nulls_bar = ax.bar(0.25, nulls, 0.5, label='NULL')
+            not_nulls_bar = ax.bar(0.75, not_nulls, 0.5, label='NOT NULL')
```
Comment (mhawryluk) [inline src/widgets/plots/null_freq_plot.py]: I would consider switching to a pie chart

File: src/widgets/unfold_widgets/preprocessing_widget.py [hunk 8]
```diff
@@ -156,17 +157,48 @@ def add_columns_to_layout(self):
...
+            self.engine.clean_data("remove")
+            self.get_data()
+
+    def remove_nulls_warning(self):
+        warning = QMessageBox()
+        warning.setIcon(QMessageBox.Warning)
+        warning.setText("Null values in set")
+        warning.setInformativeText("Some rows will be deleted cause of the null values. You can check your data or "
+                                   "estimate it. Or you might go further, letting app remove some rows automatically. "
+                                   "Continue?")
```
Comment (mhawryluk) [inline src/widgets/unfold_widgets/preprocessing_widget.py]: "This data contains some empty values. After proceeding some of the rows will be discarded. Continue?"

File: src/widgets/unfold_widgets/preprocessing_widget.py [hunk 9]
```diff
@@ -156,17 +157,48 @@ def add_columns_to_layout(self):
...
+        warning.setStandardButtons(QMessageBox.Ok | QMessageBox.Cancel)
+        warning.buttonClicked.connect(self.handle_warning_click)
+        warning.exec_()
+
+    def handle_warning_click(self, button):
+        match button.text():
+            case 'OK':
+                self.data_submitted = True
+            case 'Cancel':
+                self.data_submitted = False
```
Comment (mhawryluk) [inline src/widgets/unfold_widgets/preprocessing_widget.py]:
```suggestion
        self.data_submitted = button.text() == "OK"
```

**Comment by t0maboro** [comment]
Added some minor changes from reviews. Imo it might be a good moment to consider merging it because basic functionalities should be covered. If I omitted something important, let me know.
I plan to add the next features* in another branch, to have safeness of the stable version of that app part.
*Things, to be done: merging datasets, estimating missing data, reducing dimensions.

File: src/widgets/plots/pie_plot.py [hunk 10]
```diff
@@ -11,8 +11,13 @@ def plot(self):
...
             data_size = self.data.size
-            labels = [k if counts[k]/data_size > self.min_pie_plot_label_ratio else '' for k in counts.keys()]
-            ax.pie(counts.values(), labels=labels)
+            labels = [k for k in counts.keys() if counts[k]/data_size > self.min_pie_plot_label_ratio]
+            values = [counts[label] for label in labels]
+            values_sum = sum(values)
+            if values_sum < data_size:
+                labels.append('Other')
+                values.append(data_size-values_sum)
+            ax.pie(values, labels=labels)
```
Comment (gregori0o) [inline src/widgets/plots/pie_plot.py]: It is possible to get one blue circle with label 'Other'. I think you must add at least one label before add 'Other'.
image

File: src/widgets/plots/plot.py [hunk 11]
```diff
@@ -7,6 +7,8 @@
 class Plot(ABC):
     def __init__(self, data):
         self.data = data
+        self.max_labels_show = 10
+        self.min_pie_plot_label_ratio = 0.05
```
Comment (gregori0o) [inline src/widgets/plots/plot.py]: Do you try some smaller value of `self.min_pie_plot_label_ratio`?

File: src/widgets/plots/pie_plot.py (hunk 10 above)
Comment (mhawryluk) [inline src/widgets/plots/pie_plot.py]: I would consider excluding empty values when plotting the pie chart. I think right now they are always included as part of "Other".

File: src/widgets/plots/plot.py (hunk 11 above)
Comment (gregori0o) [reply src/widgets/plots/plot.py]: I think `0.03` value is good

**Review by gregori0o — APPROVED** [review]

**Review by mhawryluk — APPROVED** [review]
I noticed one thing worth considering. Overall awesome :)

**Pull request merged**